Change torch.library.impl to accept a device string

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111661
* #111660
* #111310
* __->__ #111659
* #111380

torch.library.impl now accepts a device string (e.g. "cpu", "cuda"). It
still accepts DispatchKey strings, but we no longer document this, because
using arbitrary DispatchKeys is more for the power users.

We map the device string to a DispatchKey and then register the impl for
said DispatchKey. A user may also specify multiple device strings at once
or specify "types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

zou3519 committed:
Change torch.library.impl to accept a device string

torch.library.impl now accepts a device string (e.g. "cpu", "cuda"). It
still accepts DispatchKey strings, but we no longer document this, because
using arbitrary DispatchKeys is more for the power users.

We map the device string to a DispatchKey and then register the impl for
said DispatchKey. A user may also specify multiple device strings at once
or specify "types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change torch.library.impl to accept a device string"


torch.library.impl now accepts a device string (e.g. "cpu", "cuda"). It
still accepts DispatchKey strings, but we no longer document this, because
using arbitrary DispatchKeys is more for the power users.

We map the device string to a DispatchKey and then register the impl for
said DispatchKey. A user may also specify multiple device strings at once
or specify "types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

[ghstack-poisoned]

soulitzer commented on torch/library.py:
```diff
@@ -351,7 +379,7 @@ def impl_abstract(qualname, func=None, *, lib=None, _stacklevel=1):
         >>>     return result
         >>>
-        >>> @torch.library.impl(lib, "custom_nonzero", "CPU")
+        >>> @torch.library.impl("mylib::custom_nonzero", "cpu")
```
With the new API, would it be best practice to omit passing lib all the time for simplicity, and rely on impl/define functions to create fragments behind the scenes? If that is the case should we accept the lib parameter at all, or are there other reasons for having it around?

soulitzer replied:
Could another way be just to have some kind of global mapping somewhere that remembers how to map ns to lib?

zou3519 replied:
> With the new API, would it be best practice to omit passing lib all the time for simplicity, and rely on impl/define functions to create fragments behind the scenes?

Yup

>  If that is the case should we accept the lib parameter at all, or are there other reasons for having it around?

This is useful for testing. I don't expect anyone to actually use the lib parameter, but I'm keeping it around as an option.

soulitzer replied:
I see, thanks.

soulitzer commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -772,6 +782,16 @@ void initDispatchBindings(PyObject* module) {
+    auto device = c10::Device(device_type);
+    TORCH_CHECK(
+        !device.has_index(),
+        "Expected device_type string to not a device index; got ",
```
```suggestion
        "Expected device_type string to not have a device index; got ",
```

soulitzer replied:
also device_type -> device type? 
since we didn't actually name the argument to impl `device_type`

zou3519 committed:
Update on "Change torch.library.impl to accept a device string"


torch.library.impl now accepts a device string (e.g. "cpu", "cuda"). It
still accepts DispatchKey strings, but we no longer document this, because
using arbitrary DispatchKeys is more for the power users.

We map the device string to a DispatchKey and then register the impl for
said DispatchKey. A user may also specify multiple device strings at once
or specify "types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

[ghstack-poisoned]
